### PR TITLE
show categories and create lesson with random 20 words

### PR DESCRIPTION
**Pull request by PhanThanhGiang91**

### CODE REVIEW

File: app/assets/javascripts/categories.coffee [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
+# All this logic will automatically be available in application.js.
+# You can use CoffeeScript in this file: http://coffeescript.org/
```
Comment (khanhkhanh) [inline app/assets/javascripts/categories.coffee]: Remove this file

File: app/assets/javascripts/lessons.coffee [hunk 2]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
+# All this logic will automatically be available in application.js.
+# You can use CoffeeScript in this file: http://coffeescript.org/
```
Comment (khanhkhanh) [inline app/assets/javascripts/lessons.coffee]: Remove this file

File: app/assets/stylesheets/categories.scss [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+.category-right{
+	height: 200px;
+	border: 1px solid rgb(194, 194, 194);
+	margin-bottom: 5px;
+	border-radius: 5px;
+}
+.title{
+	font-size: 16px;
+	margin-right: 10px;
+}
```
Comment (khanhkhanh) [inline app/assets/stylesheets/categories.scss]: Break line.

File: app/assets/stylesheets/categories.scss [hunk 4]
```diff
@@ -0,0 +1,14 @@
+// Place all the styles related to the categories controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: http://sass-lang.com/
+
+.category-right{
```
Comment (khanhkhanh) [inline app/assets/stylesheets/categories.scss]: Add a space before {

File: app/assets/stylesheets/categories.scss [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+	height: 200px;
+	border: 1px solid rgb(194, 194, 194);
+	margin-bottom: 5px;
+	border-radius: 5px;
+}
+.title{
```
Comment (khanhkhanh) [inline app/assets/stylesheets/categories.scss]: Add a space before {

File: app/assets/stylesheets/lessons.scss [hunk 6]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the lessons controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: http://sass-lang.com/
```
Comment (khanhkhanh) [inline app/assets/stylesheets/lessons.scss]: Remove this file

File: app/controllers/lessons_controller.rb [hunk 7]
```diff
@@ -0,0 +1,24 @@
+class LessonsController < ApplicationController
+  before_action :random_20_words, only: [:index]
```
Comment (khanhkhanh) [inline app/controllers/lessons_controller.rb]: Remove []

File: app/controllers/lessons_controller.rb [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+  def index
+  end
+
+  private
+
+    def random_20_words
```
Comment (khanhkhanh) [inline app/controllers/lessons_controller.rb]: Indent

File: app/controllers/lessons_controller.rb [hunk 9]
```diff
@@ -0,0 +1,24 @@
+class LessonsController < ApplicationController
+  before_action :random_20_words, only: [:index]
+
+  def index
+  end
+
```
Comment (khanhkhanh) [inline app/controllers/lessons_controller.rb]: Remove this line

File: app/helpers/categories_helper.rb [hunk 10]
```diff
@@ -0,0 +1,2 @@
+module CategoriesHelper
+end
```
Comment (khanhkhanh) [inline app/helpers/categories_helper.rb]: Remove this file

File: app/helpers/lessons_helper.rb [hunk 11]
```diff
@@ -0,0 +1,2 @@
+module LessonsHelper
+end
```
Comment (khanhkhanh) [inline app/helpers/lessons_helper.rb]: Remove this file

File: app/assets/stylesheets/categories.scss [hunk 12]
```diff
@@ -0,0 +1,11 @@
...
+	height: 200px;
+	border: 1px solid rgb(194, 194, 194);
+	margin-bottom: 5px;
+	border-radius: 5px;
+}
+
+.title {
+	font-size: 16px;
+	margin-right: 10px;
+}
```
Comment (khanhkhanh) [inline app/assets/stylesheets/categories.scss]: Indent (all the file) and lack of break line

File: app/controllers/lessons_controller.rb [hunk 13]
```diff
@@ -0,0 +1,20 @@
+class LessonsController < ApplicationController
+  before_action :random_20_words, only: :index
+  def index
```
Comment (khanhkhanh) [inline app/controllers/lessons_controller.rb]: Need a break line after before_action line

File: app/controllers/lessons_controller.rb [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+  end
+  private
```
Comment (khanhkhanh) [inline app/controllers/lessons_controller.rb]: Need a break line above private line

File: app/views/categories/index.html.erb [hunk 15]
```diff
@@ -0,0 +1,12 @@
+<% @categories.each do |category| %>
+  <div class="row">
+    <div class="col-md-6">
+      <p><strong class="title"><%= category.title %></strong> <%= category.description %></p>
+      <br>
+      <p><%= category.content %></p>
+      <br>
+      <a href="<%= lessons_path(id: category.id) %>" class="btn btn-primary pull-right">Start</a>
```
Comment (khanhkhanh) [inline app/views/categories/index.html.erb]: link_to instead of <a> tag

File: config/routes.rb [hunk 16]
```diff
@@ -4,4 +4,6 @@
   get 'about', to: 'static_pages#about', as: :about_page
```
Comment (khanhkhanh) [inline config/routes.rb]: Change ' => "

File: app/controllers/lessons_controller.rb [hunk 17]
```diff
@@ -0,0 +1,22 @@
+class LessonsController < ApplicationController
+  before_action :random_20_words, only: :index
+
+  def index
+  end
+  
+  private
+    def random_20_words
```
Comment (khanhkhanh) [inline app/controllers/lessons_controller.rb]: Indent (để def ngang hàng với private nhé)

File: app/assets/stylesheets/categories.scss [hunk 18]
```diff
@@ -0,0 +1,12 @@
...
+  height: 200px;
+  border: 1px solid rgb(194, 194, 194);
+  margin-bottom: 5px;
+  border-radius: 5px;
+}
+
+.title {
+  font-size: 16px;
+  margin-right: 10px;
+}
```
Comment (khanhkhanh) [inline app/assets/stylesheets/categories.scss]: Break line

File: app/views/categories/index.html.erb [hunk 19]
```diff
@@ -0,0 +1,12 @@
+<% @categories.each do |category| %>
+  <div class="row">
+    <div class="col-md-6">
+      <p><strong class="title"><%= category.title %></strong> <%= category.description %></p>
+      <br>
+      <p><%= category.content %></p>
+      <br>
+      <%= link_to t.('Start'), lessons_path(id: category.id), :class => "btn btn-primary pull-right" %>
```
Comment (khanhkhanh) [inline app/views/categories/index.html.erb]: t.('Start') => t(:start), and define "start" inside i18n file

File: config/routes.rb [hunk 20]
```diff
@@ -1,7 +1,9 @@
 Rails.application.routes.draw do
   root 'static_pages#home'
-  get 'help', to: 'static_pages#help', as: :help_page
-  get 'about', to: 'static_pages#about', as: :about_page
+  get 'help' => 'static_pages#help', as: :help_page
```
Comment (khanhkhanh) [inline config/routes.rb]: Change all ' to "

File: app/controllers/lessons_controller.rb [hunk 21]
```diff
@@ -0,0 +1,22 @@
...
+      if current_user.present?
+        if params[:id].present?
+          @lesson = current_user.lessons.create(category_id: params[:id].to_i)
```
Comment (khanhkhanh) [inline app/controllers/lessons_controller.rb]: Try removing ()

File: app/controllers/lessons_controller.rb [hunk 22]
```diff
@@ -0,0 +1,22 @@
...
+          Word.all.sample(20).each do |word|
+            @lesson_word = @lesson.lesson_words.create(word_id: word.id)
```
Comment (khanhkhanh) [inline app/controllers/lessons_controller.rb]: Try removing ()

File: app/controllers/lessons_controller.rb [hunk 23]
```diff
@@ -0,0 +1,22 @@
...
+      if params[:category_id].present?
+        @lesson = current_user.lessons.create category_id: params[:category_id]
+        Word.all.sample(20).each do |word|
+          @lesson_word = @lesson.lesson_words.create word_id: word.id
+        end
+      else
+        redirect_to :back
+      end
+    else
+      redirect_to '/'
```
Comment (khanhkhanh) [inline app/controllers/lessons_controller.rb]: ' => "

File: app/assets/stylesheets/categories.scss [hunk 24]
```diff
@@ -0,0 +1,12 @@
+
```
Comment (khanhkhanh) [inline app/assets/stylesheets/categories.scss]: Remove this line

File: config/locales/en.yml [hunk 25]
```diff
@@ -31,4 +31,6 @@ en:
   user: 
     sign_up: "Sign Up"
     login: "Log In"
+  categories:  
+    start: 'Start'
```
Comment (khanhkhanh) [inline config/locales/en.yml]: ' => "

File: config/routes.rb [hunk 26]
```diff
@@ -1,7 +1,9 @@
 Rails.application.routes.draw do
   root 'static_pages#home'
-  get 'help', to: 'static_pages#help', as: :help_page
-  get 'about', to: 'static_pages#about', as: :about_page
+  get 'help', to: "static_pages#help", as: :help_page
```
Comment (khanhkhanh) [inline config/routes.rb]: ' => "

File: config/routes.rb [hunk 27]
```diff
@@ -1,7 +1,9 @@
...
+  get 'about', to: "static_pages#about", as: :about_page
```
Comment (khanhkhanh) [inline config/routes.rb]: ' => "

File: app/views/categories/index.html.erb [hunk 28]
```diff
@@ -0,0 +1,12 @@
+<% @categories.each do |category| %>
+  <div class="row">
+    <div class="col-md-6">
+      <p><strong class="title"><%= category.title %></strong> <%= category.description %></p>
+      <br>
+      <p><%= category.content %></p>
+      <br>
+      <%= link_to t(:start, scope: :categories), lessons_path(category_id: category.id), :class => "btn btn-primary pull-right" %>
```
Comment (khanhkhanh) [inline app/views/categories/index.html.erb]: class: "btn btn-primary pull-right"

File: app/controllers/categories_controller.rb [hunk 29]
```diff
@@ -0,0 +1,5 @@
+class CategoriesController < ApplicationController
+  def index
+    @categories = Category.all
```
Comment (dieunb) [inline app/controllers/categories_controller.rb]: I don't see your pagination

File: app/controllers/lessons_controller.rb [hunk 30]
```diff
@@ -0,0 +1,22 @@
+class LessonsController < ApplicationController
+  before_action :random_20_words, only: :index
```
Comment (dieunb) [inline app/controllers/lessons_controller.rb]: How do you think about using call_back to random 20 words?

File: app/controllers/lessons_controller.rb [hunk 31]
```diff
@@ -0,0 +1,22 @@
...
+      if params[:category_id].present?
+        @lesson = current_user.lessons.create category_id: params[:category_id]
+        Word.all.sample(20).each do |word|
+          @lesson_word = @lesson.lesson_words.create word_id: word.id
+        end
+      else
+        redirect_to :back
+      end
+    else
+      redirect_to "/"
```
Comment (dieunb) [inline app/controllers/lessons_controller.rb]: use `root_path` instead of this

File: app/views/categories/index.html.erb [hunk 32]
```diff
@@ -0,0 +1,12 @@
+<% @categories.each do |category| %>
```
Comment (dieunb) [inline app/views/categories/index.html.erb]: I think you should try `render partial`

File: config/routes.rb [hunk 33]
```diff
@@ -1,7 +1,9 @@
 Rails.application.routes.draw do
-  root 'static_pages#home'
-  get 'help', to: 'static_pages#help', as: :help_page
-  get 'about', to: 'static_pages#about', as: :about_page
+  root "static_pages#home"
+  get "help", to: "static_pages#help", as: :help_page
+  get "about", to: "static_pages#about", as: :about_page
 
   devise_for :users
+  resources :categories
```
Comment (dieunb) [inline config/routes.rb]: should filter routes with `only` or `except`

File: config/routes.rb [hunk 34]
```diff
@@ -1,7 +1,9 @@
...
+  resources :lessons
```
Comment (dieunb) [inline config/routes.rb]: should filter routes with `only` or `except`

File: Gemfile [hunk 35]
```diff
@@ -27,6 +27,7 @@ gem 'devise'
 #Gem boostrap
 gem "bootstrap-sass",  "3.2.0.0"
 
+gem 'will_paginate', '~> 3.0.6'
```
Comment (dieunb) [inline Gemfile]: replace `'` with `"` and apply for all your code

File: app/controllers/categories_controller.rb [hunk 36]
```diff
@@ -0,0 +1,5 @@
+class CategoriesController < ApplicationController
+  def index
+    @categories = Category.all.paginate(:page => params[:page], :per_page => 3)
```
Comment (dieunb) [inline app/controllers/categories_controller.rb]:
```
Category.paginate page: params[:page], per_page: 3
```

File: app/models/lesson.rb [hunk 37]
```diff
@@ -1,5 +1,21 @@
...
+  def self.parse category_id=nil, current_user=nil
+    if current_user.present?
+      if category_id.present?
+        lesson = current_user.lessons.create category_id: category_id
+        Word.all.sample(20).each do |word|
+          lesson_word = lesson.lesson_words.create word_id: word.id
+        end
+        return lesson
+      else
+        redirect_to :back
```
Comment (dieunb) [inline app/models/lesson.rb]: I don't understand why do you push it at model.

File: app/views/categories/index.html.erb [hunk 38]
```diff
@@ -0,0 +1 @@
+<%= render partial: "categories" %>
```
Comment (dieunb) [inline app/views/categories/index.html.erb]: `render partial: "category", collection: @categories`

File: config/initializers/will_paginate.rb [hunk 39]
```diff
@@ -0,0 +1,10 @@
+if defined?(WillPaginate)
```
Comment (dieunb) [inline config/initializers/will_paginate.rb]: Is it necessary?

File: config/routes.rb [hunk 40]
```diff
@@ -1,7 +1,9 @@
...
-  root 'static_pages#home'
-  get 'help', to: 'static_pages#help', as: :help_page
-  get 'about', to: 'static_pages#about', as: :about_page
+  root "static_pages#home"
+  get "help", to: "static_pages#help", as: :help_page
+  get "about", to: "static_pages#about", as: :about_page
 
   devise_for :users
+  resources :categories, only: :index
+  get "/lesson/edit/:category_id", to: "lessons#edit", as: :edit_lesson
```
Comment (dieunb) [inline config/routes.rb]: Is it necessary?

File: app/models/lesson.rb [hunk 41]
```diff
@@ -1,5 +1,21 @@
 class Lesson < ActiveRecord::Base
   has_many :lesson_words
-  has_many :words, through :lesson_words
-  belongs_to :user  
+  has_many :words, through: :lesson_words
+  belongs_to :user
+
+  def self.parse category_id=nil, current_user=nil
+    if current_user.present?
```
Comment (dieunb) [inline app/models/lesson.rb]: You cann't call current_user in model

File: app/models/category.rb [hunk 42]
```diff
@@ -1,4 +1,7 @@
 class Category < ActiveRecord::Base
   has_many :words
   has_many :lessons
+  
+  self.per_page = 3
```
Comment (dieunb) [inline app/models/category.rb]: Is it necessary?

File: app/models/category.rb [hunk 43]
```diff
@@ -1,4 +1,7 @@
...
+  WillPaginate.per_page = 3
```
Comment (dieunb) [inline app/models/category.rb]: Is it necessary?

File: app/controllers/categories_controller.rb [hunk 44]
```diff
@@ -0,0 +1,5 @@
+class CategoriesController < ApplicationController
+  def index
+    @categories = Category.all.paginate(:page => params[:page], per_page: 3)
```
Comment (dieunb) [inline app/controllers/categories_controller.rb]: `page: params[:page], per_page: 3)`

File: app/controllers/lessons_controller.rb [hunk 45]
```diff
@@ -0,0 +1,12 @@
+class LessonsController < ApplicationController
+  def edit
+    
+  end
+
+  def create
+    lesson = Lesson.new(category_id: params[:category_id])
```
Comment (dieunb) [inline app/controllers/lessons_controller.rb]: remove `()`

File: app/models/lesson.rb [hunk 46]
```diff
@@ -1,5 +1,14 @@
 class Lesson < ActiveRecord::Base
   has_many :lesson_words
-  has_many :words, through :lesson_words
-  belongs_to :user  
+  has_many :words, through: :lesson_words
+  belongs_to :user
+  after_save :random_20_words
+
+  def random_20_words
```
Comment (dieunb) [inline app/models/lesson.rb]: private function

File: app/models/lesson.rb [hunk 47]
```diff
@@ -1,5 +1,14 @@
...
+    Word.all.sample(20).each do |word|
```
Comment (dieunb) [inline app/models/lesson.rb]: random follow with category

File: app/models/lesson.rb [hunk 48]
```diff
@@ -1,5 +1,15 @@
 class Lesson < ActiveRecord::Base
   has_many :lesson_words
-  has_many :words, through :lesson_words
-  belongs_to :user  
+  has_many :words, through: :lesson_words
+  belongs_to :user
+  after_save :random_20_words
```
Comment (chuanhtuan) [inline app/models/lesson.rb]: after_create

File: app/models/lesson.rb [hunk 49]
```diff
@@ -1,5 +1,15 @@
...
+
+  private
+  def random_20_words
+    Word.where(category_id: self.category_id).sample(20).each do |word|
```
Comment (chuanhtuan) [inline app/models/lesson.rb]: category.words.sample(20)

File: app/controllers/lessons_controller.rb [hunk 50]
```diff
@@ -0,0 +1,12 @@
+class LessonsController < ApplicationController
+  def edit
+    
+  end
+
+  def create
+    lesson = Lesson.new category_id: params[:category_id]
```
Comment (chuanhtuan) [inline app/controllers/lessons_controller.rb]: current_user.lessons.build

File: app/controllers/categories_controller.rb [hunk 51]
```diff
@@ -0,0 +1,5 @@
+class CategoriesController < ApplicationController
+  def index
+    @categories = Category.all.paginate(page: params[:page], per_page: 3)
```
Comment (chuanhtuan) [inline app/controllers/categories_controller.rb]: ()

File: app/models/lesson.rb [hunk 52]
```diff
@@ -1,5 +1,15 @@
...
   has_many :lesson_words
-  has_many :words, through :lesson_words
-  belongs_to :user  
+  has_many :words, through: :lesson_words
+  belongs_to :user
+  after_create :random_20_words
+
+  private
+  def random_20_words
+  	category = Category.find(self.category_id)
```
Comment (chuanhtuan) [inline app/models/lesson.rb]: indent

Comment (chuanhtuan) [inline app/models/lesson.rb]: ()

**Pull request merged**